<p style="text-align:center">
    <a href="https://skills.network" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo"  />
    </a>
</p>


# **Impute Missing Values**


Estimated time needed: **30** minutes


In this lab, you will practice essential data wrangling techniques using the Stack Overflow survey dataset. The primary focus is on handling missing data and ensuring data quality. You will:

- **Load the Data:** Import the dataset into a DataFrame using the pandas library.

- **Clean the Data:** Identify and remove duplicate entries to maintain data integrity.

- **Handle Missing Values:** Detect missing values, impute them with appropriate strategies, and verify the imputation to create a complete and reliable dataset for analysis.

This lab equips you with the skills to effectively preprocess and clean real-world datasets, a crucial step in any data analysis project.


## Objectives


In this lab, you will perform the following:


-   Identify missing values in the dataset.

-   Apply techniques to impute missing values in the dataset.
  
-   Use suitable techniques to normalize data in the dataset.


-----


#### Install needed library


In [2]:
!pip install pandas

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.9/10.9 MB 188.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.6/16.6 MB 165.0 MB/s eta 0:00:00


### Step 1: Import Required Libraries


In [3]:
import pandas as pd

### Step 2: Load the Dataset Into a Dataframe


#### **Read Data**
<p>
The functions below will download the dataset into your browser:
</p>


In [4]:
file_path = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/n01PQ9pSmiRX6520flujwQ/survey-data.csv"
df = pd.read_csv(file_path)

# Display the first few rows to ensure it loaded correctly
print(df.head())

   ResponseId                      MainBranch                 Age  \
0           1  I am a developer by profession  Under 18 years old   
1           2  I am a developer by profession     35-44 years old   
2           3  I am a developer by profession     45-54 years old   
3           4           I am learning to code     18-24 years old   
4           5  I am a developer by profession     18-24 years old   

            Employment RemoteWork   Check  \
0  Employed, full-time     Remote  Apples   
1  Employed, full-time     Remote  Apples   
2  Employed, full-time     Remote  Apples   
3   Student, full-time        NaN  Apples   
4   Student, full-time        NaN  Apples   

                                    CodingActivities  \
0                                              Hobby   
1  Hobby;Contribute to open-source projects;Other...   
2  Hobby;Contribute to open-source projects;Other...   
3                                                NaN   
4                                 

### Step 3. Finding and Removing Duplicates
##### Task 1: Identify duplicate rows in the dataset.


In [5]:
## Write your code here
# 1. Count the number of duplicate rows
df.duplicated().sum()

# 2. Display the first few duplicate rows
df[df.duplicated()].head()

Empty DataFrame
Columns: [ResponseId, MainBranch, Age, Employment, RemoteWork, Check, CodingActivities, EdLevel, LearnCode, LearnCodeOnline, TechDoc, YearsCode, YearsCodePro, DevType, OrgSize, PurchaseInfluence, BuyNewTool, BuildvsBuy, TechEndorse, Country, Currency, CompTotal, LanguageHaveWorkedWith, LanguageWantToWorkWith, LanguageAdmired, DatabaseHaveWorkedWith, DatabaseWantToWorkWith, DatabaseAdmired, PlatformHaveWorkedWith, PlatformWantToWorkWith, PlatformAdmired, WebframeHaveWorkedWith, WebframeWantToWorkWith, WebframeAdmired, EmbeddedHaveWorkedWith, EmbeddedWantToWorkWith, EmbeddedAdmired, MiscTechHaveWorkedWith, MiscTechWantToWorkWith, MiscTechAdmired, ToolsTechHaveWorkedWith, ToolsTechWantToWorkWith, ToolsTechAdmired, NEWCollabToolsHaveWorkedWith, NEWCollabToolsWantToWorkWith, NEWCollabToolsAdmired, OpSysPersonal use, OpSysProfessional use, OfficeStackAsyncHaveWorkedWith, OfficeStackAsyncWantToWorkWith, OfficeStackAsyncAdmired, OfficeStackSyncHaveWorkedWith, OfficeStackSyncWantToWorkWith, OfficeStackSyncAdmired, AISearchDevHaveWorkedWith, AISearchDevWantToWorkWith, AISearchDevAdmired, NEWSOSites, SOVisitFreq, SOAccount, SOPartFreq, SOHow, SOComm, AISelect, AISent, AIBen, AIAcc, AIComplex, AIToolCurrently Using, AIToolInterested in Using, AIToolNot interested in Using, AINextMuch more integrated, AINextNo change, AINextMore integrated, AINextLess integrated, AINextMuch less integrated, AIThreat, AIEthics, AIChallenges, TBranch, ICorPM, WorkExp, Knowledge_1, Knowledge_2, Knowledge_3, Knowledge_4, Knowledge_5, Knowledge_6, Knowledge_7, Knowledge_8, Knowledge_9, Frequency_1, Frequency_2, Frequency_3, TimeSearching, TimeAnswering, Frustration, ProfessionalTech, ProfessionalCloud, ProfessionalQuestion, ...]
Index: []

[0 rows x 114 columns]

##### Task 2: Remove the duplicate rows from the dataframe.



In [6]:
## Write your code here
# Remove duplicates (even though Task 1 showed there are none)
df = df.drop_duplicates()

# Verify removal by counting duplicates again
duplicates_after = df.duplicated().sum()

print("Number of duplicate rows after removal:", duplicates_after)

Number of duplicate rows after removal: 0


In [8]:
# # Forcing pandas to left-align the index
# pd.set_option('display.unicode.east_asian_width', True)
# pd.set_option('display.unicode.ambiguous_as_wide', True)
# pd.set_option('display.colheader_justify', 'left')


### Step 4: Finding Missing Values
##### Task 3: Find the missing values for all columns.


In [7]:
## Write your code here
missing_values = df.isnull().sum()
print("Missing values per column:\n", missing_values)

Missing values per column:
 ResponseId                 0
MainBranch                 0
Age                        0
Employment                 0
RemoteWork             10631
                       ...  
JobSatPoints_11        35992
SurveyLength            9255
SurveyEase              9199
ConvertedCompYearly    42002
JobSat                 36311
Length: 114, dtype: int64


##### Task 4: Find out how many rows are missing in the column RemoteWork.


In [9]:
## Write your code here
missing_employment = df['RemoteWork'].isnull().sum()
print("Number of missing rows in 'RemoteWork':", missing_employment)


Number of missing rows in 'RemoteWork': 10631


### Step 5. Imputing Missing Values
##### Task 5: Find the value counts for the column RemoteWork.


In [10]:
# Show the distribution of non-missing values
print("\nValue counts for 'RemoteWork':\n") 
print(df['RemoteWork'].value_counts())



Value counts for 'RemoteWork':

RemoteWork
Hybrid (some remote, some in-person)    23015
Remote                                  20831
In-person                               10960
Name: count, dtype: int64


##### Task 6: Identify the most frequent (majority) value in the RemoteWork column.



In [13]:
# Find the most frequent value (Mode)
# .mode() returns a Series, so we take the first element [0]
most_frequent_value = df['RemoteWork'].mode()[0]
print("\nMost frequent value selected for imputation:", most_frequent_value)
# print(f"The most frequent value is: {most_frequent_value}")



Most frequent value selected for imputation: Hybrid (some remote, some in-person)


##### Task 7: Impute (replace) all the empty rows in the column RemoteWork with the majority value.



In [14]:
## Write your code here
# Impute missing values with the most frequent value (no inplace)
df['RemoteWork'] = df['RemoteWork'].fillna(most_frequent_value)

# Verify that missing values have been handled
print("\nMissing values in 'RemoteWork' after imputation:", df['RemoteWork'].isnull().sum())
# print(f"Remaining missing values: {df['RemoteWork'].isnull().sum()}")



Missing values in 'RemoteWork' after imputation: 0
Remaining missing values: 0


##### Task 8: Check for any compensation-related columns and describe their distribution.



In [17]:
# Identify compensation-related columns by keyword
comp_cols = [col for col in df.columns if 'Comp' in col or 'Pay' in col or 'Salary' in col]

print("Compensation-related columns found:\n", comp_cols)

# Describe each compensation column
for col in comp_cols:
    print(f"\nDistribution summary for {col}:\n")
    print(df[col].describe())


Compensation-related columns found:
 ['CompTotal', 'AIComplex', 'ConvertedCompYearly']

Distribution summary for CompTotal:

count     3.374000e+04
mean     2.963841e+145
std      5.444117e+147
min       0.000000e+00
25%       6.000000e+04
50%       1.100000e+05
75%       2.500000e+05
max      1.000000e+150
Name: CompTotal, dtype: float64

Distribution summary for AIComplex:

count                                             37021
unique                                                5
top       Good, but not great at handling complex tasks
freq                                              12102
Name: AIComplex, dtype: object

Distribution summary for ConvertedCompYearly:

count    2.343500e+04
mean     8.615529e+04
std      1.867570e+05
min      1.000000e+00
25%      3.271200e+04
50%      6.500000e+04
75%      1.079715e+05
max      1.625660e+07
Name: ConvertedCompYearly, dtype: float64


### Compensation patterns you just uncovered

- **CompTotal** is clearly corrupted or contains extreme outliers.  
  The mean (`2.96e+145`) and max (`1e+150`) are astronomically large—far beyond any real salary. This usually happens when:
  - the column contains mixed types (strings accidentally converted to floats)
  - respondents entered unrealistic values
  - the dataset includes placeholder or synthetic values

- **AIComplex** is not a numeric compensation field.  
  It’s a categorical column describing how respondents perceive AI’s ability to handle complex tasks.

- **ConvertedCompYearly** is the cleanest and most realistic compensation field.  
  Its distribution looks typical for global salary data:
  - median around **65,000**
  - long right tail up to **16 million** (likely outliers but still within numeric range)
  - standard deviation large due to global variation

### A concise interpretation you can include in your lab

- **CompTotal** contains extreme outliers or corrupted values, making it unreliable for analysis without cleaning or winsorization.
- **AIComplex** is not a compensation variable and should be excluded from salary analysis.
- **ConvertedCompYearly** is the most usable compensation field, showing a realistic salary distribution with a long right tail.


In [19]:
# import matplotlib.pyplot as plt
# import seaborn as sns

# plt.figure(figsize=(14, 6))

# # Histogram
# plt.subplot(1, 2, 1)
# sns.histplot(df['ConvertedCompYearly'], bins=50, kde=True, color='teal')
# plt.title("ConvertedCompYearly Distribution")
# plt.xlabel("Yearly Compensation")
# plt.ylabel("Frequency")

# # Boxplot
# plt.subplot(1, 2, 2)
# sns.boxplot(x=df['ConvertedCompYearly'], color='teal')
# plt.title("ConvertedCompYearly Boxplot (Outliers Visible)")
# plt.xlabel("Yearly Compensation")

# plt.tight_layout()
# plt.show()


In [21]:
# # Histogram
# df['ConvertedCompYearly'].plot(kind='hist', bins=50, figsize=(12, 5), color='teal', title='ConvertedCompYearly Distribution')
# plt.xlabel("Yearly Compensation")
# plt.show()

# # Boxplot
# df['ConvertedCompYearly'].plot(kind='box', figsize=(10, 4), color='teal', title='ConvertedCompYearly Boxplot')
# plt.show()


In [24]:
# # 1. List all columns related to compensation to identify them
# comp_cols = [col for col in df.columns if 'Comp' in col]
# print("Compensation-related columns found:", comp_cols)

# # 2. Describe the distribution (Statistical Summary)
# # This shows count, mean, std, min, 25%, 50% (median), 75%, and max
# print("\nStatistical Summary of Compensation:")
# print(df['ConvertedCompYearly'].describe())

# # 3. Visualize the Distribution
# plt.figure(figsize=(12, 6))

# # Use a Histogram with a Kernel Density Estimate (KDE) line
# sns.histplot(df['ConvertedCompYearly'], bins=50, kde=True, color='forestgreen')

# plt.title('Distribution of Annual Compensation (ConvertedCompYearly)', fontsize=15)
# plt.xlabel('Annual Compensation (USD)', fontsize=12)
# plt.ylabel('Frequency', fontsize=12)

# # Optional: Adding a vertical line for the Median
# median_val = df['ConvertedCompYearly'].median()
# plt.axvline(median_val, color='red', linestyle='--', label=f'Median: {median_val:,.0f}')
# plt.legend()

# plt.show()

### Summary 


**In this lab, you focused on imputing missing values in the dataset.**

- Use the <code>pandas.read_csv()</code> function to load a dataset from a CSV file into a DataFrame.

- Download the dataset if it's not available online and specify the correct file path.



<!--
## Change Log
|Date (YYYY-MM-DD)|Version|Changed By|Change Description|
|-|-|-|-|
|2024-11-05|1.3|Madhusudhan Moole|Updated lab|
|2024-10-29|1.2|Madhusudhan Moole|Updated lab|
|2024-09-27|1.1|Madhusudhan Moole|Updated lab|
|2024-09-26|1.0|Raghul Ramesh|Created lab|
--!>


Copyright © IBM Corporation. All rights reserved.
